File: Assignment_45X/Assignment_45.3.cpp
#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

class SinglyLL
{
    private:
        PNODE first;
        int iCount;

    public:

        SinglyLL()
        {
            cout<<"Object of SinglyLL gets created"<<"\n";
            first = NULL;
            iCount = 0;
        }

    void InsertFirst(int no)
    {
        PNODE newn = NULL;

        newn = new NODE;

        newn->data = no;
        newn->next = NULL;

        if(first == NULL)
        {
            first = newn;
        }
        else
        {
            newn->next = first;
            first = newn;
        }
    }

    void Display()
    {
        PNODE temp = NULL;

        temp = first;

        while(temp != NULL)
        {
            cout<<"| "<<temp->data<<" |->";
            temp = temp->next;
        }

        cout<<"NULL\n";
    }

    int LastOcc(int iNo)
    {
        PNODE temp = NULL;
        int iPos = 1;
        int Lastpos = 0;

        temp = first;

        while(temp != NULL)
        {
            if((temp->data) == iNo)
            {
               Lastpos = iPos;
            }
            
            temp = temp->next;
            iPos++;
            
        }
        return Lastpos;
    }
    
};

int main()
{
    SinglyLL obj;
    int iRet = 0;

    obj.InsertFirst(51);
    obj.InsertFirst(40);
    obj.InsertFirst(31);
    obj.InsertFirst(20);
    obj.InsertFirst(11);
    obj.InsertFirst(31);

    obj.Display();

    iRet = obj. LastOcc(31);
     
    cout<<"Position Of First Occurance is: "<<iRet;

    return 0;
}

File: Assignment_45XX/Assignment_45_3.java
class node
{
    public int data;
    public node next;
    
    public node(int no)
    {
        data = no;
        next = null;
    }
}

class SinglyLL
{
    private node first;
    private int iCount;

    public SinglyLL() 
    {
        System.out.println("Object Of SinglyLL gets created");
        first = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {
        node newn  = null;

        newn = new node(no);

        if(first == null)
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first = newn;
        }
    }
    public void Display()
    {
        node temp = null;

        temp = first;

        while(temp != null)
        {
            System.out.print("| "+temp.data+" |->");
            temp = temp.next;
        }

        System.out.println("null\n");
    }

    public int LastOcc(int iNo)
    {
        node temp = null;

        temp = first;
        int lastpos = 0;
        
        int iPos = 1;

        while(temp != null)
        {
            if((temp.data) == iNo)
            {
               lastpos = iPos;
                 
            }
            iPos++;
            temp = temp.next;
        }
    
        return lastpos;
    }

    
}

class Assignment_45_3
{
    public static void main(String Arr[])
    {
        SinglyLL obj;

        int iRet = 0;

        obj = new SinglyLL();

        obj.InsertFirst(51);
        obj.InsertFirst(41);
        obj.InsertFirst(31);
        obj.InsertFirst(10);
        obj.InsertFirst(10);
        obj.InsertFirst(24);
        obj.InsertFirst(10);
        obj.InsertFirst(4);

        obj.Display();

        iRet = obj.LastOcc(10);

        System.out.println("First Occurance Position is :"+iRet);

        obj = null;

        System.gc();
    }
}